src/Process/Runner/Parallel/ParallelProcessRunner.php

```diff
@@ -49,69 +48,95 @@
  * @internal
  * @final
  *
- * This ProcessManager is a simple wrapper to enable parallel processing using Symfony Process component
+ * This ProcessManager is an elaborate wrapper to enable parallel processing using Symfony Process component
  */
 class ParallelProcessRunner
 {
-    private $eventDispatcher;
+    /**
+     * If it takes 100000 ms for a process to finish, and 5000 ms to make it,
+     * then we can make as much as 20 processes while we wait. But let's not
+     * get greedy and settle on a smaller number to make sure we're not stuck
+     * making processes where we should be starting them.
+     */
+    private const MUTATOR_TO_PROCESS_RATIO = 10;
+
+    private $processHandler;
 
     /**
-     * @var MutantProcess[]
+     * @var ProcessBearer[]
      */
-    private $currentProcesses = [];
+    private $runningProcesses = [];
 
-    public function __construct(EventDispatcher $eventDispatcher)
+    public function __construct(Closure $processHandler)
     {
-        $this->eventDispatcher = $eventDispatcher;
+        $this->processHandler = $processHandler;
     }
 
     /**
-     * @param MutantProcess[] $processes
+     * @param ProcessBearer[] $processes
      *
      * @throws RuntimeException
      * @throws LogicException
      */
     public function run(iterable $processes, int $threadCount, int $poll = 1000): void
     {
+        /*
+         * It takes about 100000 ms for a mutated process to finish, where it takes
+         * about 5000 ms to make it. Therefore instead of just waiting we can produce
+         * new processes so that when a process or several finish, we would have
+         * additional jobs on hand, without a need to wait for them to be created.
+         *
+         * For our purposes we need to make sure we only see one process only once.
+         */
+        $generator = self::toGenerator($processes);
```

Comment:
```suggestion
        $generator = iter\toIter($processes);
```

Reply:
[I'll be rewindable.](https://github.com/nikic/iter/blob/master/src/iter.php#L1024) Generators are not. I want to be very explicit about this here.

`NoRewindIterator` is another can of worms if you ask.